dali/operators/reader/fits_reader_op.h

```diff
@@ -0,0 +1,86 @@
+// Copyright (c) 2020-2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_OPERATORS_READER_FITS_READER_OP_H_
+#define DALI_OPERATORS_READER_FITS_READER_OP_H_
+
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "dali/operators/reader/loader/fits_loader.h"
+#include "dali/operators/reader/reader_op.h"
+#include "dali/pipeline/operator/arg_helper.h"
+#include "dali/util/crop_window.h"
+
+namespace dali {
+
+template <typename Backend, typename Target>
+class FitsReader : public DataReader<Backend, Target> {
+ public:
+  explicit FitsReader(const OpSpec& spec) : DataReader<Backend, Target>(spec) {}
+
+  bool CanInferOutputs() const override {
+    return true;
+  }
+
+  USE_READER_OPERATOR_MEMBERS(Backend, Target);
+  using DataReader<Backend, Target>::GetCurrBatchSize;
+  using DataReader<Backend, Target>::GetSample;
+  using Operator<Backend>::spec_;
+
+  bool SetupImpl(std::vector<OutputDesc>& output_desc, const Workspace& ws) override {
+    // If necessary start prefetching thread and wait for a consumable batch
+    DataReader<Backend, Target>::SetupImpl(output_desc, ws);
+    int num_outputs = ws.NumOutput();
+    int num_samples = GetCurrBatchSize();  // samples here are synonymous with files
+
+    output_desc.resize(num_outputs);
+    const auto& sample_0 = GetSample(0);
+
+    for (int output_idx = 0; output_idx < num_outputs; output_idx++) {
+      int ndim = sample_0.data[output_idx].shape().sample_dim();
+      output_desc[output_idx].shape = TensorListShape<>(num_samples, ndim);
+    }
+
+    for (int sample_idx = 0; sample_idx < num_samples; sample_idx++) {
+      auto& sample = GetSample(sample_idx);
+      for (int output_idx = 0; output_idx < num_outputs; output_idx++) {
+        output_desc[output_idx].shape.set_tensor_shape(sample_idx, sample.data[output_idx].shape());
+        output_desc[output_idx].type = sample.data[output_idx].type();
```

Comment:
Do you (or one of the base classes?) validate the `dtypes`  against what we actually get in the header?